### ISSUE TITLE
[Feature Request] Postponed evaluation of PEP 563-compliant type aliases

### ISSUE THREAD

**Issue by patrick-kidger** [beartype/beartype#222]
```python
from __future__ import annotations

from beartype import beartype
from typing import Generic, TypeVar

T = TypeVar("T")

class Foo(Generic[T]):
    pass

@beartype
def f(x: Bar):
    pass

Bar = Foo[int]
f(Foo())
# beartype.roar.BeartypeCallHintForwardRefException: Forward reference "__main__.Qux" referent __main__.Foo[int] not class.
```

**Comment by leycec** [comment]
This... *is a weird one.* But it's also a fascinating one. Static type-checkers operate outside time and space itself. So, it's trivial for *them* to statically parse the `Qux.py` script and deduce that `Bar` is (of course) a type alias for `Foo[int]`. Order of operations is mostly irrelevant at static analysis time.

Sadly, @beartype operates inside time and space. At the time that the `@beartype` decorator is applied to the function `f()`, the type hint `Bar` has yet to be defined. Despite that, `@beartype` still has to wrap `f()` with *some* sort of dynamically generated runtime type-checking that validates `x` as an instance of the undefined type hint `Bar`. Right? Gotta do somethin'.

Currently, `@beartype` assumes that all stringified type hints are forward references to *classes* that have yet to be defined. But `Foo[int]` isn't a class; it's a subscripted generic, exactly as you say. Thankfully, the class underlying a subscripted generic can be easily extracted from that generic via the `__origin__` dunder attribute: e.g.,

```python
>>> Foo[int].__origin__
<class '__main__.Qux.Foo'>
``` 

Clearly, this calls for desperate measures. When scarce volunteer time and my beautiful wife permit, <sup>*...so, by 2047 at the earliest*</sup> @beartype should be generalized to do the following when confronted with any forward reference at runtime type-checking time:

* If that reference resolves to a class, we're done. Praise be to moral turpitude.
* If that reference resolves to a subscripted generic, reduce that generic to its origin class (i.e., `generic.__origin__`).
* Else, raise an exception.

## So Why Not Just Do That Already, Then?

Interestingly, this feature request intersects with *another* feature request you "recently" submitted. <sup>...okay, it was at the exact same time</sup> I speak ominously of #217, AKA "Postponed evaluation of stringified type hints."

To support #217, @beartype will need to `eval()` any stringified type hint that is *not* a simple Python identifier (i.e., any type hint that is a string containing any punctuation). If doing so fails with an `AttributeError` at decoration time (e.g., due to the type hint actually being a forward reference to a subscripted generic that has yet to be declared **as in the example above**), @beartype will need to:

1. Silently squelch that exception.
2. Defer the lookup of that hint until runtime type-checking time.

Pretty sure this is all doable, but... the sweat beading my forehead is real. :cold_sweat:

**Comment by leycec** [comment]
@beartype summons the ultimate destructive force to deal with this long-standing issue: apocalyptic anime exploding girl, deal with this one too!

![arise ugh](https://media.tenor.com/V6k3KbrycpsAAAAC/anime-explosion.gif)

**Comment by leycec** [comment]
**Resolved by e82a1f7142...** but *mostly* by a18d9a5a5da, which this issue *mostly* turned out to be a duplicate of. I shall now summon [my younger self circa last Friday night](https://github.com/beartype/beartype/issues/226#issuecomment-1694160566), who personally witnessed [the tragic chain of events that ultimately lead to @beartype's redemption and this final resolution](https://github.com/beartype/beartype/issues/226#issuecomment-1694160566):

## Younger Self Witnesses All, Now Tells Sad Story (Only on GitHub)

**omg! OMG! OMGGGGG!!!** @TeamSpen210: [Your brilliant five-sentence abstract on the nature of Python vis-a-vis forward references](https://github.com/beartype/beartype/issues/226#issuecomment-1484045900) *actually* worked! It *actually* worked, bro! Punch the air like you just don't care, 'cause that's what we're doin'.

![do it like this](https://media.tenor.com/hV03QzfXrtEAAAAC/hatsune-priconne.gif)

Naturally, your "simpler approach" was horrifying. My once-beautific bald face is crinkled with wrinkles, my twitchy eye is even more twitchy, and our cat won't stop pacing the room. I spent the entirety of my vacation doing this – and I don't even *need* this in production code. It just became a titanic struggle of @leycec versus @beartype versus Python: a three-way untelevised cage match that could only end with one man hurtling fifty feet onto another man lying prostrate on a table perched on top of another man curled in the fetal position while bleeding out live on GitHub.

![mankind is sad](https://media.tenor.com/t-Tw6ma_jxIAAAAC/undertaker-mankind.gif)
<sup>...dat the best u can do, python?</sup>

![wat](https://media.tenor.com/T1E389Up3RgAAAAC/jeff-hardy-whisper-in-the-wind.gif)
<sup>*wat is even happening here*</sup>

![uh gods](https://media.tenor.com/3U1flZRg0f8AAAAC/adam-cole-johnny-gargano.gif)
<sup>When your masterplan only ends in tragedy for everybody.</sup>

Seriously. I've got five hundred-line comments devsplaining single one-liners. I've got 100KB of incomprehensible pure-Python isolated to four new submodules sprawled across a new subpackage devoted exclusively to inchoate madness. I've got abstract base classes, abstract metaclasses, dynamic type factories, dunder methods on sunder methods on abstract cached properties. I've got friggin' metaclass `__getattr__()` methods performing deferred module attribute importation via dynamic class attribute lookups of attributes that don't actually physically exist anywhere. I've got two whole `dict.__missing__()` implementations that don't even really have anything to do with one another. *This is inchoate madness.*

I don't even know how any of this works anymore. I've already jettisoned all knowledge pertaining to this topic so I can just sleep tonight without fitfully crying out in the night for my childhood stuffed octopus. I wish to awaken from the real-world nightmare I have unwittingly exposed myself to. It turns out *Bloodborne* is more a state of mind rather than a mere video game.

![who knew](https://media.tenor.com/-AMHQLsgQXcAAAAd/bloodborne-werewolf.gif)
<sup>Python. *It do be like that.*</sup>

Forward references: do not go there, lest this animated reenactment of @leycec's life corrupt your once-pure essence.

I'm pretty sure (but not certain) that I may have inadvertently summoned the Eldritch She-God Shub-Niggurath, The Black Goat of the Woods with a Thousand Young, during the resolution of this feature request. I apologize, humanity. I'm so sorry! I didn't *mean* to open that portal. It just opened of its own accord while the guard cat – a fat Maine Coon named Nekali, of course – was indolently passed out on the shag carpeting.
